hypershift-operator/controllers/scheduler/dedicated_request_serving_nodes.go

```diff
@@ -285,3 +290,530 @@ func (r *DedicatedServingComponentScheduler) Reconcile(ctx context.Context, req
+	// and pair labels.
+	var goalNodes []corev1.Node
+	for _, node := range dedicatedNodesForHC {
+		if node.Labels[hyperv1.NodeSizeLabel] == desiredSize && node.Labels[OSDFleetManagerPairedNodesLabel] == pairLabel {
+			goalNodes = append(goalNodes, node)
+		}
+	}
+
+	// Find any nodes that are in the same fleet manager group and have the right size
+	// but are not labeled with the hosted cluster label. Ensure that these nodes are labeled
+	// and tainted with the hosted cluster label. This can happen if not all nodes were labeled/tainted
+	// when they were initially selected.
+	if pairLabel != "" {
+		var needClusterLabel []corev1.Node
+		for _, node := range availableNodes {
+			if node.Labels[hyperv1.NodeSizeLabel] == desiredSize && node.Labels[OSDFleetManagerPairedNodesLabel] == pairLabel {
+				needClusterLabel = append(needClusterLabel, node)
+			}
+		}
+		if len(needClusterLabel) > 0 {
+			for _, node := range needClusterLabel {
+				if err := r.ensureHostedClusterLabelAndTaint(ctx, hc, &node); err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+
+	goalNodeZones := sets.New[string]()
+	for _, node := range goalNodes {
+		if zone := node.Labels["topology.kubernetes.io/zone"]; zone != "" {
+			goalNodeZones.Insert(zone)
+		}
+	}
+
+	if len(goalNodeZones) > 1 {
+		// If we have enough nodes, update the hosted cluster.
+		if err := r.updateHostedCluster(ctx, hc, desiredSize, &config, goalNodes); err != nil {
+			return ctrl.Result{}, err
+		}
+		// Ensure we don't have a placeholder deployment, since we have nodes
+		if err := r.deletePlaceholderDeployment(ctx, hc); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// If there isn't a current pair label, then we can select from available nodes selected by placeholders.
+	if pairLabel == "" {
+		sizeConfig := sizeConfiguration(&config, desiredSize)
+		if sizeConfig == nil {
+			return ctrl.Result{}, fmt.Errorf("could not find size configuration for size %s", desiredSize)
+		}
+
+		// If placeholders are present, use those
+		if sizeConfig.Management != nil && sizeConfig.Management.Placeholders > 0 {
+			placeholderPods, candidateNodes, err := r.nodesFromPlaceholders(ctx, desiredSize)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to get nodes from placeholders: %w", err)
+			}
+			if len(candidateNodes) > 0 {
+				for _, node := range candidateNodes {
+					if err := r.ensureHostedClusterLabelAndTaint(ctx, hc, &node); err != nil {
+						return ctrl.Result{}, err
+					}
+				}
+				for _, pod := range placeholderPods {
+					if err := r.Delete(ctx, &pod); err != nil {
+						return ctrl.Result{}, fmt.Errorf("failed to delete placeholder pod: %w", err)
+					}
+				}
+				return ctrl.Result{Requeue: true}, nil
+			}
+		}
+	}
+
+	// Create a deployment to ensure nodes of the right size are created
+	nodesNeeded := 2 - len(goalNodes)
```

Comment:
I assume `len(goalNodes)` can't be greater than 2, right?
should we use min just to be sure?
```golang
nodesNeeded := min(2 - len(goalNodes), 0)
````